scipy/stats/stats.py

```diff
@@ -4226,6 +4232,395 @@ def weightedtau(x, y, rank=True, weigher=None, additive=True):
     return WeightedTauResult(_weightedrankedtau(x, y, rank, weigher, additive), np.nan)
 
 
+# FROM MGCPY: https://github.com/neurodata/mgcpy
+
+def _perm_test(x, y, stat, ind_test, compute_distance, reps=1000):
+    r"""
+    Helper function that calculates the p-value. See below for uses.
+
+    Parameters
+    ----------
+    x, y : array_like
+        `x` and `y` are be :math:`n \times p` and :math:`n \times q` data
+        matrices.
+    stat : float
+        The sample MGC test statistic within :math:`[-1, 1]`.
+    ind_test : callable
+        The independence test statistic calculation function used.
+    compute_distance : callable
+        A function that computes the distance or similarity among the samples
+        within each data matrix. Set to `None` if `x` and `y` are already
+        distance.
+    reps : int, optional
+        The number of replications used to estimate the null when using the
+        permutation test. The default is 1000 repelications.
+
+    Returns
+    -------
+    stat : float
+        The sample test p-value.
+    """
+    # generate null distribution and p-value
+    pvalue = 0
+    null_dist = np.zeros(reps)
+    for i in range(reps):
+        # randomly permute one of the data matrices
+        permy = np.random.permutation(y)
+
+        # calculate permuted stats and MGC map
+        perm_stat, _ = ind_test(x, permy, compute_distance)
+        null_dist[i] = perm_stat
+
+        # calculate p-value
+        pvalue += ((perm_stat >= stat) * (1/reps))
+
+    # correct for a p_value of 0. This is because, with bootstrapping
+    # permutations, a value of 0 is incorrect
+    if pvalue == 0:
+        pvalue = 1 / reps
+
+    return pvalue, null_dist
+
+
+def _euclidean_dist(x):
+    return squareform(pdist(x, metric='euclidean'))
```

Comment:
Isn't this `cdist(x, x)`? If defining a private function to calculate Euclidean distance is really justified, maybe a brief comment is in order, since we have so many ways of doing this already

Reply:
The reason that `squareform(pdist(x, metric='euclidean'))` was used was so that pairwise distances are always going to be calculated for this input. If I was to use `cdist(x, x)` as the default, that would require `compute_distance` to be of the form `compute_distance(x, x)` rather than just `compute_distance(x)`.

I could probably just change `squareform(pdist(x, metric='euclidean'))` to `cdist(x, x)` if that works?

Reply:
https://github.com/scipy/scipy/pull/10524/commits/390e52a9eab7e4c29c0dea72cd0814fe26445fac